migration-engine/core/src/migration/datamodel_calculator.rs

```diff
@@ -1,219 +1,423 @@
-            MigrationStep::CreateField(x) => apply_create_field(&mut result, x),
-            MigrationStep::CreateIndex(x) => apply_create_index(&mut result, x),
-            MigrationStep::DeleteIndex(x) => apply_delete_index(&mut result, x),
-            MigrationStep::UpdateIndex(x) => apply_update_index(&mut result, x),
-        });
-        result
+    fn infer(&self, current: &SchemaAst, steps: &[MigrationStep]) -> SchemaAst {
+        let cloned: SchemaAst = current.clone();
+        apply(cloned, steps).unwrap()
+    }
+}
+
+fn apply(mut schema: SchemaAst, steps: &[MigrationStep]) -> crate::Result<SchemaAst> {
+    for step in steps {
+        apply_step(&mut schema, step);
+    }
+
+    Ok(schema)
+}
+
+fn apply_step(datamodel: &mut ast::SchemaAst, step: &MigrationStep) {
+    match step {
+        MigrationStep::CreateEnum(create_enum) => apply_create_enum(datamodel, create_enum),
+        MigrationStep::UpdateEnum(update_enum) => apply_update_enum(datamodel, update_enum),
+        MigrationStep::DeleteEnum(delete_enum) => apply_delete_enum(datamodel, delete_enum),
+        MigrationStep::CreateModel(create_model) => apply_create_model(datamodel, create_model),
+        MigrationStep::UpdateModel(update_model) => apply_update_model(datamodel, update_model),
+        MigrationStep::DeleteModel(delete_model) => apply_delete_model(datamodel, delete_model),
+        MigrationStep::CreateField(create_field) => apply_create_field(datamodel, create_field),
+        MigrationStep::UpdateField(update_field) => apply_update_field(datamodel, update_field),
+        MigrationStep::DeleteField(delete_field) => apply_delete_field(datamodel, delete_field),
+        MigrationStep::CreateDirective(create_directive) => apply_create_directive(datamodel, create_directive),
+        MigrationStep::DeleteDirective(delete_directive) => apply_delete_directive(datamodel, delete_directive),
+        MigrationStep::CreateDirectiveArgument(create_directive_argument) => {
+            apply_create_directive_argument(datamodel, create_directive_argument)
+        }
+        MigrationStep::DeleteDirectiveArgument(delete_directive_argument) => {
+            apply_delete_directive_argument(datamodel, delete_directive_argument)
+        }
+        MigrationStep::UpdateDirectiveArgument(update_directive_argument) => {
+            apply_update_directive_argument(datamodel, update_directive_argument)
+        }
     }
 }
 
-fn apply_delete_model(data_model: &mut Datamodel, step: &DeleteModel) {
-    if !data_model.has_model(&step.name) {
-        panic!(
-            "The model {} does not exist in this Datamodel. It is not possible to delete it.",
-            &step.name
-        )
+fn apply_create_enum(datamodel: &mut ast::SchemaAst, step: &steps::CreateEnum) {
+    let steps::CreateEnum { r#enum: name, values } = step;
+
+    if let Some(_) = datamodel.find_enum(&name) {
+        Err::<(), _>(format_err!(
```

Comment:
I like simpler syntax usually :/ this ended up like that because I converted old panics. I think we could just do "formatted panic", here we would still get the stack trace.

Reply:
Addressed in b725d0cb8c63c407a5ae6a896bf8f74b582d2eae